vault/resource_terraform_cloud_secret_creds.go

```diff
@@ -0,0 +1,148 @@
+package vault
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/vault/api"
+)
+
+func terraformCloudSecretCredsResource() *schema.Resource {
+	return &schema.Resource{
+		Create: createTerraformCloudSecretCredsResource,
+		Read:   readTerraformCloudSecretCredsResource,
+		Update: updateTerraformCloudSecretCredsResource,
+		Delete: deleteTerraformCloudSecretCredsResource,
+		Schema: map[string]*schema.Schema{
+			"backend": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Terraform Cloud secret backend to generate tokens from.",
+			},
+			"role": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the role.",
+			},
+			"token": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Terraform Token provided by the Vault backend",
+				Sensitive:   true,
+			},
+			"token_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "ID of the Terraform Token provided",
+			},
+			"organization": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Optional:    true,
+				Description: "Name of the Terraform Cloud or Enterprise organization",
+			},
+			"team_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Optional:    true,
+				Description: "ID of the Terraform Cloud or Enterprise team under organization (e.g., settings/teams/team-xxxxxxxxxxxxx)",
+			},
+		},
+	}
+}
+
+func createTerraformCloudSecretCredsResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	backend := d.Get("backend").(string)
+	role := d.Get("role").(string)
+	path := fmt.Sprintf("%s/creds/%s", backend, role)
+
+	secret, err := client.Logical().Read(path)
+	if err != nil {
+		return fmt.Errorf("error reading from Vault: %s", err)
+	}
+	log.Printf("[DEBUG] Read %q from Vault", path)
+
+	if secret == nil {
+		return fmt.Errorf("no role found at %q", path)
+	}
+
+	token := secret.Data["token"].(string)
+	if token == "" {
+		return fmt.Errorf("token is not set in response")
+	}
+	tokenId := secret.Data["token_id"].(string)
+
+	organization := secret.Data["organization"]
+	teamId := secret.Data["team_id"]
+
+	if secret.LeaseID != "" {
+		d.SetId(secret.LeaseID)
+	} else {
+		d.SetId(tokenId)
+	}
+	d.Set("token", token)
+	d.Set("token_id", tokenId)
+	d.Set("organization", organization)
+	d.Set("team_id", teamId)
+
+	return nil
+}
+
+func deleteTerraformCloudSecretCredsResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	id := d.Id()
+
+	err := client.Sys().Revoke(id)
```

Comment:
I don't believe this actually revokes team or org tokens, as their IDs are not lease IDs. It turns out this endpoint returns a `204 no content` regardless of whatever lease id you give it

```console
$ curl -i -s --header "X-Vault-Token: $(vault print token)" -X PUT -d '{"lease_id":"ihavethehighground"}' http://127.0.0.1:8200/v1/sys/leases/revoke 
HTTP/1.1 204 No Content
Cache-Control: no-store
Content-Type: application/json
Date: Fri, 12 Mar 2021 21:28:19 GMT
```

If you try to revoke a team or org token, it appears to succeed, but instead the token remains (still visible on app.terraform.io)

Reply:
My understanding is that we don't want the token to be destroyed on the site? IIRC, the token is created externally and then referenced through ENV vars. Because it's created outside of Terraform, my instinct was to leave it in place.

Reply:
Follow-up: we discussed this synchronously and decided to leave as is. Resolving this.